File: lib_mvm/src/main/java/com/mirkowu/lib_mvm/util/RxLife.java
package com.mirkowu.lib_mvm.util;

import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleOwner;

import com.mirkowu.lib_mvm.core.ILifecycleBinder;

import autodispose2.AutoDispose;
import autodispose2.AutoDisposeConverter;
import autodispose2.androidx.lifecycle.AndroidLifecycleScopeProvider;

/***
 * 绑定生命周期
 */
public class RxLife {

    public static <R> AutoDisposeConverter<R> bindLifecycle(ILifecycleBinder binder) {
        return bindLifecycle(binder.getLifecycleOwner(), binder.bindLifecycleUntil());
    }

    public static <R> AutoDisposeConverter<R> bindLifecycle(LifecycleOwner owner, Lifecycle.Event untilEvent) {
        return AutoDispose.autoDisposable(AndroidLifecycleScopeProvider.from(owner, untilEvent));
    }

    public static <R> AutoDisposeConverter<R> bindLifecycleAuto(LifecycleOwner owner) {
        return AutoDispose.autoDisposable(AndroidLifecycleScopeProvider.from(owner));
    }

}

File: lib_base/src/main/java/com/mirkowu/lib_base/util/RxLife.kt
package com.mirkowu.lib_base.util

import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import autodispose2.AutoDispose
import autodispose2.ObservableSubscribeProxy
import autodispose2.androidx.lifecycle.AndroidLifecycleScopeProvider
import com.mirkowu.lib_base.view.ILifecycleBinder
import io.reactivex.rxjava3.core.Observable

/**
 * Lifecycle自动解绑
 * eg.
 * Observable.bindLifecycle(lifecycle)
 *
 */

fun <R : Any> Observable<R>.bindLifecycle(binder: ILifecycleBinder): ObservableSubscribeProxy<R> {
    return bindLifecycle(binder.lifecycleOwner, binder.bindLifecycleUntil())
}

fun <R : Any> Observable<R>.bindLifecycle(owner: LifecycleOwner, untilEvent: Lifecycle.Event): ObservableSubscribeProxy<R> {
    return to(AutoDispose.autoDisposable(AndroidLifecycleScopeProvider.from(owner, untilEvent)))
}

fun <R : Any> Observable<R>.bindLifecycleAuto(owner: LifecycleOwner): ObservableSubscribeProxy<R> {
    return to(AutoDispose.autoDisposable(AndroidLifecycleScopeProvider.from(owner)))
}
